microproxy/layer/proxy/socks.py

```diff
@@ -33,120 +33,108 @@ def process_and_return_context(self):
             if isinstance(_event, GreetingRequest):
-                event = self.handle_greeting_request(_event)
-                data = self.socks_conn.send(event)
-                yield self.context.src_stream.write(data)
+                yield self.handle_greeting_request(_event)
             elif isinstance(_event, Request):
-                error, event = yield self.handle_request(_event)
-                if event:
-                    data = self.socks_conn.send(event)
-                    yield self.context.src_stream.write(data)
-                if error:
-                    raise error
+                dest_stream, host, port = yield self.handle_request_and_create_destination(_event)
+                self.context.dest_stream = dest_stream
+                self.context.host = host
+                self.context.port = port
                 break
+            else:
+                raise NotImplementedError("not handling with {0}".format(_event))
 
-        self.context.src_stream.pause()
         raise gen.Return(self.context)
 
+    @gen.coroutine
     def handle_greeting_request(self, event):
         if not AUTH_TYPE["NO_AUTH"] in event.methods:
-            response_event = GreetingResponse(
-                socks_version, AUTH_TYPE["NO_SUPPORT_AUTH_METHOD"])
+            yield self.send_event_to_src_conn(GreetingResponse(
+                socks_version, AUTH_TYPE["NO_SUPPORT_AUTH_METHOD"]))
         else:
-            response_event = GreetingResponse(
-                socks_version, AUTH_TYPE["NO_AUTH"])
-        return response_event
+            yield self.send_event_to_src_conn(GreetingResponse(
+                socks_version, AUTH_TYPE["NO_AUTH"]))
 
     @gen.coroutine
-    def handle_request(self, event):
+    def handle_request_and_create_destination(self, event):
+        """Handle the socks requet from source
+        Create destination connection
 
+        Returns:
+            tuple: (dest_stream, host, port)
+        """
         if event.cmd != REQ_COMMAND["CONNECT"]:
             logger.debug("Unsupport connect type")
-            error = ProtocolError("Unsupport bind type")
-            response_event = Response(
+            self.send_event_to_src_conn(Response(
                 socks_version, RESP_STATUS["COMMAND_NOT_SUPPORTED"],
-                event.atyp, event.addr, event.port)
-            raise gen.Return((error, response_event))
+                event.atyp, event.addr, event.port), raise_exception=False)
+            raise ProtocolError("Unsupport bind type")
 
-        host = event.addr
-        port = event.port
-        dest_stream = None
         try:
-            dest_stream = yield self.create_dest_stream((host, port))
-
-        except Exception as e:
-            err, response_event = self.handle_connection_error(
-                e, event, dest_stream)
-
+            dest_stream = yield self.create_dest_stream((event.addr, event.port))
+        except gen.TimeoutError as te:
```

Comment:
is te and se really necessary? it seems the gen.Timeout and StreamClosedError are pretty self explain.